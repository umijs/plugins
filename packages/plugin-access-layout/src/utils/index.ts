import fs from 'fs';

export function getScriptPath(filepath: string): string {
  let realFilePath: string = '';
  if (fs.existsSync(`${filepath}.ts`)) {
    realFilePath = `${filepath}.ts`;
  } else if (fs.existsSync(`${filepath}.js`)) {
    realFilePath = `${filepath}.js`;
  }
  return realFilePath;
}

export function checkIfHasDefaultExporting(filepath: string): boolean {
  const scriptPath = getScriptPath(filepath);
  if (!scriptPath) {
    return false;
  }

  const fileContent = fs.readFileSync(scriptPath, 'utf8');
  const validationRegExp = /(export\s*default)|(exports\.default)|(module.exports[\s\S]*default)|(module.exports[\s\n]*=)/m;

  return validationRegExp.test(fileContent);
}
