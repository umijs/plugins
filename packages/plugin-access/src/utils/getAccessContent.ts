export default function() {
  return `\
import React, { useContext } from 'react';
import AccessContext, { AccessInstance as AccessInstanceType } from './context';

export type AccessInstance = AccessInstanceType;

export const useAccess = () => {
  const access = useContext(AccessContext);

  return access;
};

export interface AccessProps {
  accessible: boolean;
  fallback?: React.ReactNode;
}

export const Access: React.FC<AccessProps> = props => {
  const { accessible, fallback, children } = props;

  if (process.env.NODE_ENV === 'development' && typeof accessible === 'function') {
    console.warn(
      '[plugin-access]: provided "accessible" prop is a function named "' +
        (accessible as Function).name +
        '" instead of a boolean, maybe you need check it.',
    );
  }

  return <>{accessible ? children : fallback}</>;
};
`;
}
