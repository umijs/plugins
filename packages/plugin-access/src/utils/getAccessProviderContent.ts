import { utils } from 'umi';
import { join } from 'path';

export default function(util: typeof utils) {
  return `\
import React, { useMemo } from 'react';
import { IRoute } from 'umi';
import { useModel } from '../core/umiExports';
import accessFactory from '../../access';
import AccessContext, { AccessInstance } from './context';
import { traverseModifyRoutes } from '${util.winPath(
    join(__dirname, '..', 'utils', 'runtimeUtil'),
  )}';

type Routes = IRoute[];

interface Props {
  routes: Routes;
  children: React.ReactNode;
}

const AccessProvider: React.FC<Props> = props => {
  if (typeof useModel !== 'function') {
    throw new Error('[plugin-access]: useModel is not a function, @umijs/plugin-initial-state is needed.')
  }

  const { children } = props;
  const { initialState } = useModel('@@initialState');

  const access: AccessInstance = useMemo(() => accessFactory(initialState as any), [initialState]);

  if (process.env.NODE_ENV === 'development' && (access === undefined || access === null)) {
    console.warn('[plugin-access]: the access instance created by access.ts(js) is nullish, maybe you need check it.');
  }

  return React.createElement(
    AccessContext.Provider,
    { value: access },
    React.cloneElement(children, {
      ...children.props,
      routes:traverseModifyRoutes(props.routes, access)
    }),
  );
};

export default AccessProvider;
`;
}
