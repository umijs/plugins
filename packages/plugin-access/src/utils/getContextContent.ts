export default function() {
  return `\
import React from 'react';
import accessFactory from '@/access';

export type AccessInstance = ReturnType<typeof accessFactory>;

const AccessContext = React.createContext<AccessInstance>(null!);

export default AccessContext;
`;
}
