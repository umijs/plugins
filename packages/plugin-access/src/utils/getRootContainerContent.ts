export default function() {
  return `\
import React from 'react';
import AccessProvider from './AccessProvider';

export function rootContainer(container: React.ReactNode, { routes }) {
  return React.createElement(AccessProvider, { routes }, container);
}
`;
}
