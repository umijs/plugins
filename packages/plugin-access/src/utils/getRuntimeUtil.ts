// This file is for runtime, not the compile time.
export default (strictMode: boolean = false) =>
  `import { IRoute } from 'umi';

  type ChildrenList = IRoute[];

  const oldChildrenPropsName = 'routes';
  
  export function traverseModifyRoutes(childrenList: ChildrenList, access: any): ChildrenList {
    const resultChildrenList: ChildrenList = []
      .concat(childrenList as any)
      .map((resultRoute: IRoute) => {
        const childList = resultRoute.children || resultRoute[oldChildrenPropsName];
        if (childList && childList?.length) {
          return {
            ...resultRoute,
            children: childList?.map((route: any) => ({ ...route })),
            // return new route to routes.
            [oldChildrenPropsName]: childList?.map((route: any) => ({ ...route })),
          };
        }
        return resultRoute;
      });
  
    return resultChildrenList.map((currentRoute) => {
      let currentRouteAccessible =
        typeof currentRoute.unaccessible === 'boolean' ? !currentRoute.unaccessible : true;

      if(currentRoute && ${strictMode}){
        currentRoute.unaccessible = true;
      }

      // 判断路由是否有权限的具体代码
      if (currentRoute && currentRoute.access) {
        if (typeof currentRoute.access !== 'string') {
          throw new Error(
            '[plugin-access]: "access" field set in "' +
              currentRoute.path +
              '" route should be a string.',
          );
        }
        const accessProp = access[currentRoute.access];
        // 如果是方法需要执行以下
        if (typeof accessProp === 'function') {
          currentRouteAccessible = accessProp(currentRoute);
        } else if (typeof accessProp === 'boolean') {
          // 不是方法就直接 copy
          currentRouteAccessible = accessProp;
        }
        currentRoute.unaccessible = !currentRouteAccessible;
      }
      const childList = currentRoute.children || currentRoute[oldChildrenPropsName];
      // 筛选子路由
      if (childList && Array.isArray(childList) && childList.length) {
        if (!Array.isArray(childList)) {
          return currentRoute;
        }
        // 父亲没权限，理论上每个孩子都没权限
        // 可能有打平 的事情发生，所以都执行一下
        childList.forEach((childRoute) => {
          childRoute.unaccessible = !currentRouteAccessible;
        });
        const finallyChildList = traverseModifyRoutes(childList, access);
  
        // 如果每个子节点都没有权限，那么自己也属于没有权限
        const isAllChildChildrenUnaccessible =
          Array.isArray(finallyChildList) && finallyChildList.every((route) => route.unaccessible);
  
        if (!currentRoute.unaccessible && isAllChildChildrenUnaccessible) {
          currentRoute.unaccessible = true;
        }
        if (finallyChildList && finallyChildList?.length > 0) {
          return {
            ...currentRoute,
            children: finallyChildList,
            [oldChildrenPropsName]: finallyChildList,
          };
        }
        delete currentRoute.routes;
        delete currentRoute.children;
      }
  
      return currentRoute;
    });
}`;
