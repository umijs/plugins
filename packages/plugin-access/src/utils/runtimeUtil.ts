// This file is for runtime, not the compile time.
import { IRoute } from 'umi';

type Routes = IRoute[];

export function traverseModifyRoutes(routes: Routes, access: any) {
  const resultRoutes: Routes = []
    .concat(routes as any)
    .map((resultRoute: IRoute) => {
      const { routes } = resultRoute;
      return {
        ...resultRoute,
        // return new route to routes.
        routes: routes ? routes.map((route: any) => ({ ...route })) : routes,
      };
    });
  const notHandledRoutes: Routes = [];

  notHandledRoutes.push(...resultRoutes);

  for (let i = 0; i < notHandledRoutes.length; i++) {
    const currentRoute = notHandledRoutes[i];
    let currentRouteAccessible =
      typeof currentRoute.unAccessible === 'boolean'
        ? !currentRoute.unAccessible
        : true;
    if (currentRoute && currentRoute.access) {
      if (typeof currentRoute.access !== 'string') {
        throw new Error(
          '[plugin-access]: "access" field set in "' +
            currentRoute.path +
            '" route should be a string.',
        );
      }
      const accessProp = access[currentRoute.access];
      if (typeof accessProp === 'function') {
        currentRouteAccessible = accessProp(currentRoute);
      } else if (typeof accessProp === 'boolean') {
        currentRouteAccessible = accessProp;
      }
      currentRoute.unAccessible = !currentRouteAccessible;
    }

    if (currentRoute.routes || currentRoute.childRoutes) {
      const childRoutes: Routes =
        currentRoute.routes || currentRoute.childRoutes;
      if (!Array.isArray(childRoutes)) {
        continue;
      }
      childRoutes.forEach(childRoute => {
        childRoute.unAccessible = !currentRouteAccessible;
      }); // Default inherit from parent route
      notHandledRoutes.push(...childRoutes);
    }
  }

  // Make parent route unAccessible if child routes exist and all of child routes are unAccessible
  for (let i = 0; i < notHandledRoutes.length; i++) {
    const currentRoute = notHandledRoutes[i];
    const childRoutes: Routes = currentRoute.routes || currentRoute.childRoutes;
    const isAllChildRoutesUnaccessible =
      Array.isArray(childRoutes) &&
      childRoutes.every(route => route.unAccessible);
    if (!currentRoute.unAccessible && isAllChildRoutesUnaccessible) {
      currentRoute.unAccessible = true;
    }
  }

  return resultRoutes;
}
