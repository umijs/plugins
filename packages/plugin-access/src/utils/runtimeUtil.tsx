// This file is for runtime, not the compile time.
import { IRoute } from 'umi';

type Routes = IRoute[];

export function traverseModifyRoutes(routes: Routes, access: any): Routes {
  const resultRoutes: Routes = []
    .concat(routes as any)
    .map((resultRoute: IRoute) => {
      const { routes } = resultRoute;
      if (routes && routes?.map) {
        return {
          ...resultRoute,
          // return new route to routes.
          routes: routes?.map((route: any) => ({ ...route })),
        };
      }
      return resultRoute;
    });

  return resultRoutes.map((currentRoute) => {
    let currentRouteAccessible =
      typeof currentRoute.unaccessible === 'boolean'
        ? !currentRoute.unaccessible
        : true;

    // 判断路由是否有权限的具体代码
    if (currentRoute && currentRoute.access) {
      if (typeof currentRoute.access !== 'string') {
        throw new Error(
          '[plugin-access]: "access" field set in "' +
            currentRoute.path +
            '" route should be a string.',
        );
      }
      const accessProp = access[currentRoute.access];
      // 如果是方法需要执行以下
      if (typeof accessProp === 'function') {
        currentRouteAccessible = accessProp(currentRoute);
      } else if (typeof accessProp === 'boolean') {
        // 不是方法就直接 copy
        currentRouteAccessible = accessProp;
      }
      currentRoute.unaccessible = !currentRouteAccessible;
    }

    // 筛选子路由
    if (currentRoute.routes || currentRoute.childRoutes) {
      const childRoutes: Routes =
        currentRoute.routes || currentRoute.childRoutes;

      if (!Array.isArray(childRoutes)) {
        return currentRoute;
      }
      // 父亲没权限，理论上每个孩子都没权限
      // 可能有打平 的事情发生，所以都执行一下
      childRoutes.forEach((childRoute) => {
        childRoute.unaccessible = !currentRouteAccessible;
      });
      const finallyChildRoute = traverseModifyRoutes(childRoutes, access);

      // 如果每个子节点都没有权限，那么自己也属于没有权限
      const isAllChildRoutesUnaccessible =
        Array.isArray(finallyChildRoute) &&
        finallyChildRoute.every((route) => route.unaccessible);

      if (!currentRoute.unaccessible && isAllChildRoutesUnaccessible) {
        currentRoute.unaccessible = true;
      }
      if (finallyChildRoute && finallyChildRoute?.length > 0) {
        return {
          ...currentRoute,
          routes: finallyChildRoute,
        };
      }
      delete currentRoute.routes;
    }

    return currentRoute;
  });
}
