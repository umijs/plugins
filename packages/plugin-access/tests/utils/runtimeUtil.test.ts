import { IRoute } from 'umi';
import { traverseModifyRoutes } from '../../src/utils/runtimeUtil';

let routes: IRoute[] = [];
const access = {
  canReadHomepage: () => true,
  canReadAdminPage: () => false,
  canReadAdminList: () => true,
  canReadAdminDetail: () => false,
  canReadUser: () => true,
  canReadUserDetail: () => true,
  canReadOrder: () => true,
  canReadOrderAudit: () => false,
  canReadOrderCancellation: () => false,
  canReadAbout: true,
};

describe('TraverseModifyRoutes', () => {
  beforeEach(() => {
    routes = [
      {
        path: '/homepage',
        access: 'canReadHomepage',
        routes: null as any,
      },
      {
        path: '/admin',
        access: 'canReadAdminPage',
        routes: [
          {
            path: '/adminList',
            access: 'canReadAdminList',
          },
          {
            path: '/adminDetail',
            access: 'canReadAdminDetail',
          },
        ],
      },
      {
        path: '/user',
        access: 'canReadUser',
        routes: [
          {
            path: '/userDetail',
            access: 'canReadUserDetail',
          },
        ],
      },
      {
        path: '/order',
        access: 'canReadOrder',
        routes: [
          {
            path: '/orderAudit',
            access: 'canReadOrderAudit',
          },
          {
            path: '/orderCancellation',
            access: 'canReadOrderCancellation',
          },
        ],
      },
      {
        path: '/about',
        access: 'canReadAbout',
      },
    ];
  });

  it('should get expected accessible result', () => {
    const result = traverseModifyRoutes(routes, access);
    expect(result).toMatchSnapshot();
  });

  it('test no use code', () => {
    const result = traverseModifyRoutes(
      [
        {
          path: '/homepage',
          access: 'canReadHomepage',
          routes: {} as any,
        },
        {
          path: '/admin',
          access: 'canReadAdminPage',
          routes: [],
        },
        {
          path: '/order',
          access: 'canReadUser',
          routes: [
            {
              path: '/orderAudit',
              access: 'canReadOrderAudit',
            },
            {
              path: '/orderCancellation',
              access: 'canReadOrderAudit',
            },
          ],
        },
      ],
      access,
    );
    expect(result).toMatchSnapshot();
  });

  it('should throw error if access of arbitrary route is not a string', () => {
    routes[0].access = () => {};
    expect(() => {
      traverseModifyRoutes(routes, access);
    }).toThrowError();
  });
});
