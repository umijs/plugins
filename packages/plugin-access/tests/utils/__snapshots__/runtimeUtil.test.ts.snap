// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraverseModifyRoutes should get expected accessible result 1`] = `
Array [
  Object {
    "access": "canReadHomepage",
    "path": "/homepage",
    "routes": null,
    "unaccessible": false,
  },
  Object {
    "access": "canReadAdminPage",
    "path": "/admin",
    "routes": Array [
      Object {
        "access": "canReadAdminList",
        "path": "/adminList",
        "unaccessible": false,
      },
      Object {
        "access": "canReadAdminDetail",
        "path": "/adminDetail",
        "unaccessible": true,
      },
    ],
    "unaccessible": true,
  },
  Object {
    "access": "canReadUser",
    "path": "/user",
    "routes": Array [
      Object {
        "access": "canReadUserDetail",
        "path": "/userDetail",
        "unaccessible": false,
      },
    ],
    "unaccessible": false,
  },
  Object {
    "access": "canReadOrder",
    "path": "/order",
    "routes": Array [
      Object {
        "access": "canReadOrderAudit",
        "path": "/orderAudit",
        "unaccessible": true,
      },
      Object {
        "access": "canReadOrderCancellation",
        "path": "/orderCancellation",
        "unaccessible": true,
      },
    ],
    "unaccessible": true,
  },
  Object {
    "access": "canReadAbout",
    "path": "/about",
    "unaccessible": false,
  },
]
`;

exports[`TraverseModifyRoutes test no use code 1`] = `
Array [
  Object {
    "access": "canReadHomepage",
    "path": "/homepage",
    "routes": Object {},
    "unaccessible": false,
  },
  Object {
    "access": "canReadAdminPage",
    "path": "/admin",
    "unaccessible": true,
  },
  Object {
    "access": "canReadUser",
    "path": "/order",
    "routes": Array [
      Object {
        "access": "canReadOrderAudit",
        "path": "/orderAudit",
        "unaccessible": true,
      },
      Object {
        "access": "canReadOrderAudit",
        "path": "/orderCancellation",
        "unaccessible": true,
      },
    ],
    "unaccessible": true,
  },
]
`;
